common/lib/util_sys.c

```diff
@@ -83,113 +75,124 @@ __weak int pal_submit_12v_cycle_slot()
+	ipmi_msg *me_msg = (ipmi_msg *)malloc(sizeof(ipmi_msg));
+	uint8_t *data = (uint8_t *)malloc(5 * sizeof(uint8_t));
+	if ((me_msg == NULL) || (data == NULL)) {
+		printf("[%s] Failed to allocate memory\n", __func__);
+		goto cleanup;
 	}
-
-	free(me_msg);
-	return;
-}
-
-void ME_enter_recovery()
-{
-	ipmi_msg *me_msg;
-	ipmb_error status;
-	uint8_t retry = 3;
-
-	for (int i = 0; i < retry; i++) {
-		set_ME_FW_mode(ME_FW_RECOVERY);
-		k_msleep(2000);
-
-		me_msg = (ipmi_msg *)malloc(sizeof(ipmi_msg));
-		if (me_msg == NULL) {
-			printf("ME recovery msg alloc fail\n");
-			k_msleep(10);
+	ipmb_error ret;
+	uint16_t data_len = 0;
+	uint8_t seq_source = 0xFF;
+	int i = 0, retry = 3;
+
+	for (i = 0; i < retry; i++) {
+		seq_source = 0xFF;
+		data_len = 4;
+		data[0] = INTEL_IANA & 0xFF;
+		data[1] = (INTEL_IANA >> 8) & 0xFF;
+		data[2] = (INTEL_IANA >> 16) & 0xFF;
+		data[3] = me_fw_mode;
+		*me_msg = construct_ipmi_message(seq_source, NETFN_NM_REQ,
+						 CMD_OEM_NM_FORCE_ME_RECOVERY, SELF, ME_IPMB,
+						 data_len, data);
+		ret = ipmb_read(me_msg, IPMB_inf_index_map[me_msg->InF_target]);
+		if (ret != IPMB_ERROR_SUCCESS) {
+			printf("Failed to set ME frimware mode to 0x%x, ret: 0x%x\n", me_fw_mode,
+			       ret);
 			continue;
 		}
 
-		me_msg->seq_source = 0xFF;
-		me_msg->netfn = NETFN_APP_REQ;
-		me_msg->cmd = CMD_APP_GET_SELFTEST_RESULTS;
-		me_msg->InF_source = SELF;
-		me_msg->InF_target = ME_IPMB;
-		me_msg->data_len = 0;
-		status = ipmb_read(me_msg, IPMB_inf_index_map[me_msg->InF_target]);
-		if (status == IPMB_ERROR_SUCCESS) {
-			if ((me_msg->data_len == 2) && (me_msg->data[0] == 0x81) &&
-			    (me_msg->data[1] == 0x02)) {
-				free(me_msg);
+		switch (me_fw_mode) {
+		case ME_FW_RECOVERY:
+			k_msleep(2000);
+			break;
+		case ME_FW_RESTORE:
+			k_msleep(8000);
+			break;
+		default:
+			k_msleep(1000);
+			break;
+		}
+
+		seq_source = 0xFF;
+		data_len = 0;
+		*me_msg = construct_ipmi_message(seq_source, NETFN_APP_REQ,
+						 CMD_APP_GET_SELFTEST_RESULTS, SELF, ME_IPMB,
+						 data_len, NULL);
+		ret = ipmb_read(me_msg, IPMB_inf_index_map[me_msg->InF_target]);
+		if (ret == IPMB_ERROR_SUCCESS) {
+			switch (me_fw_mode) {
+			case ME_FW_RECOVERY:
+				if ((me_msg->data_len == 2) &&
+				    (me_msg->data[0] == FIRMWARE_ENTERED_RECOVERY_MODE) &&
+				    (me_msg->data[1] == RECOVERY_MODE_CAUSE_IPMI_COMMAND)) {
+					goto cleanup;
+				}
+				break;
+			case ME_FW_RESTORE:
+				if ((me_msg->data_len == 2) && (me_msg->data[0] == NO_ERROR) &&
+				    (me_msg->data[1] == 0x00)) {
+					goto cleanup;
+				}
+				break;
+			default:
 				break;
 			}
 		} else {
-			printk("ME restore get selftest fail with status: %x\n", status);
-		}
-		if (me_msg != NULL) {
-			free(me_msg);
+			printf("Failed to get ME self test result after setting 0x%x mode, ret: 0x%x\n",
+			       me_fw_mode, ret);
 		}
 	}
+	if (i == retry) {
+		printf("Failed to set ME frimware mode to 0x%x, retry time: %d\n", me_fw_mode,
```

Comment:
ME firmware

Reply:
We have fixed this and done the forced update.
Please review it. Thanks.